Rust/src/anomaly_score_visitor.rs

```diff
@@ -0,0 +1,196 @@
+        }
+    }
+
+    /// Update the anomaly score from an internal node.
+    ///
+    /// After visiting a leaf node in [`Self::accept_leaf`] the remaining nodes
+    /// encountered while traversing up the tree to the root are internal
+    /// nodes. The main piece of information used by this algorithm is the
+    /// bounding box at these nodes.
+    pub fn accept(&mut self, node: &Internal<T>, depth: T) {
+        if self.point_inside_box { return; }
+
+        let separation_probability = self.separation_probability(node.bounding_box());
+        if separation_probability <= Zero::zero() {
+            self.point_inside_box = true;
+            return;
+        }
+
+        let one: T = One::one();
+        self.anomaly_score = separation_probability * score_unseen(depth) +
+            (one - separation_probability) * self.anomaly_score;
+    }
+
+    /// Normalize and return the anomaly score computed by the visitor.
+    ///
+    /// The anomaly score computed by the visiting process, via
+    /// [`Self::accept_leaf()`] and [`Self::accept()`], is normalized using the
+    /// mass of the tree before returning. This is so that the resulting anomaly
+    /// score is independent of the number of samples in the tree.
+    pub fn get_result(&self) -> T {
+        normalize_score(self.anomaly_score, self.tree.mass())
+    }
+
+    /// Returns the probability that the point to score and the input bounding
+    /// box are separated by a random cut.
+    fn separation_probability(&mut self, bounding_box: &BoundingBox<T>) -> T {
+        let mut new_range_sum: T = Zero::zero();
+        let mut range_diff_sum: T = Zero::zero();
+        let min_values = bounding_box.min_values();
+        let max_values = bounding_box.max_values();
+
+        for i in 0..bounding_box.dimensions() {
+            let mut min_value = min_values[i];
+            let mut max_value = max_values[i];
+            let old_range = max_value - min_value;
+
+            if !self.coordinate_inside_box[i] {
+                if max_value < self.point_to_score[i] {
+                    max_value = self.point_to_score[i]
+                } else if min_value > self.point_to_score[i] {
+                    min_value = self.point_to_score[i];
+                } else {
+                    // in this case the current coordinate lies within the width
+                    // of the bounding box in this dimension. we update with
+                    // old range and indicate that this coordinate is now inside
+                    new_range_sum = new_range_sum + old_range;
+                    self.coordinate_inside_box[i] = true;
+                    continue;
+                }
+
+                let new_range = max_value - min_value;
+                new_range_sum = new_range_sum + new_range;
+                range_diff_sum = range_diff_sum + new_range - old_range;
+            } else {
+                new_range_sum = new_range_sum + old_range;
```

Comment:
Any reason not to use new_range_sum += old_range?

Reply:
The `Float` trait, which I use to genericize over `f32` and `f64` only implements `Add` and not `AddAssign`. I wanted to minimize the number of dependent traits for the input type and managed to reduce it to `Float + Sum`. The `Sum` trait means you can call `.sum()` on a collection of that type. This requirement mostly is due to making `BoundingBox` easier to implement.

In Rust nightly you can define `trait aliases`. See [this comment](https://github.com/aws/random-cut-forest-by-aws/pull/173#issuecomment-830291529) for context and info.

Reply:
Anyway, the solution to this comment is either:

1. keep the "verbose" accumulation
2. include `AddAssign` in the trait bounds and use `+=`

I don't know what the cons are to 2. Are compilers smart enough to accumulate, here?

Reply:
I think the verbose accumulation is fine